Hyrule isn't using bytecode files

Hello!

As in the title; I tried using the master versions as well, and the issue persists. Removing `hyrule` imports and requires from my module, while keeping all other modules, sped up the import of my module.

Sorry if this isn't the place for this, as I don't know whether this counts as a bug or not.

Thank you kindly for the help!

Okay, while I can disable the following test, is there any way to fix it?
```
==================================== ERRORS ====================================
____________________ ERROR collecting tests/test_slicing.hy ____________________
ImportError while importing test module '/tmp/nix-build-python3.10-hyrule-0.1.drv-0/source/tests/test_slicing.hy'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/nix/store/zgv6h355ssr2dxydxx4hiagn2sdx9356-python3-3.10.4/lib/python3.10/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests/test_slicing.hy:3: in <module>
    (import
E   ModuleNotFoundError: No module named 'hy.lex'
=========================== short test summary info ============================
ERROR tests/test_slicing.hy
!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
=============================== 1 error in 3.73s ===============================
error: builder for '/nix/store/prb2rybzrn6gwkvblkjb7vkrccdpzsc2-python3.10-hyrule-0.1.drv' failed with exit code 2;
       last 10 log lines:
       > Traceback:
       > /nix/store/zgv6h355ssr2dxydxx4hiagn2sdx9356-python3-3.10.4/lib/python3.10/importlib/__init__.py:126: in import_module
       >     return _bootstrap._gcd_import(name[level:], package, level)
       > tests/test_slicing.hy:3: in <module>
       >     (import
       > E   ModuleNotFoundError: No module named 'hy.lex'
       > =========================== short test summary info ============================
       > ERROR tests/test_slicing.hy
       > !!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
       > =============================== 1 error in 3.73s ===============================
       For full logs, run 'nix log /nix/store/prb2rybzrn6gwkvblkjb7vkrccdpzsc2-python3.10-hyrule-0.1.drv'.
error: 1 dependencies of derivation '/nix/store/0y5dnicnv389yz061izb5dv3rh47xn8w-python3.10-oreo-1.0.0.0.drv' failed to build
```

Otherwise, @scauligi, it works! In case you use `nix` as well, I am pasting `shell.nix` file below, with the `hy[rule]` master and your branch:
```nix
with builtins; let
    pkgs = import (fetchGit { url = "https://github.com/shadowrylander/nixpkgs"; ref = "j"; }) {};
    hy = pkgs.python310Packages.hy.overridePythonAttrs (prev: {
      src = fetchGit { url = "https://github.com/scauligi/hy"; ref = "the_wheel_deal"; };
      # src = fetchGit { url = "https://github.com/hylang/hy"; ref = "master"; };
      postPatch = ''substituteInPlace setup.py --replace "\"funcparserlib ~= 1.0\"," ""'';
      disabledTests = [ "test_bin" "test_hy2py" ];
    });
    hyrule = pkgs.python310Packages.hyrule.overridePythonAttrs (prev: {
      src = fetchGit { url = "https://github.com/scauligi/hyrule"; ref = "precompile_hy"; };
      # src = fetchGit { url = "https://github.com/hylang/hyrule"; ref = "master"; };
      postPatch = ''substituteInPlace setup.py --replace "'hy @ git+https://github.com/scauligi/hy@the_wheel_deal#egg=hy'," ""'';
      propagatedBuildInputs = [ hy ];
      disabledTestPaths = [ "tests/test_slicing.hy" ];
    });
in with pkgs; mkShell {
    buildInputs = with python310Packages; [ (oreo.overridePythonAttrs (prev: {
        src = ./.;
        propagatedBuildInputs = (filter (p: ! elem p.pname [ "hy" "hyrule" ]) prev.propagatedBuildInputs) ++ [ hy hyrule ];
    })) ];
    # shellHook = "hy tests.hy; exit";
}
```

--- a/.gitignore
+++ b/.gitignore
@@ -3,3 +3,4 @@
 /dist/
 __pycache__/
 /docs/_build
+/.eggs

--- a/NEWS.rst
+++ b/NEWS.rst
@@ -25,6 +25,7 @@ Bug Fixes
   failing to match a certain pattern, instead of sometimes returning
   `None` and sometimes raising an exception.
 * `#%` reader macro now parses `%i` names from dotted symbols.
+* Hyrule now pre-compiles .hy files during setup/installation.
 
 0.1 (released 2022-01-09)
 ==============================

--- a/setup.py
+++ b/setup.py
@@ -1,18 +1,41 @@
 #!/usr/bin/env python
 
 import setuptools
+from setuptools.command.install import install
 
 with open('README.rst', 'r') as o:
     long_description = o.read()
 
+
+class install(install):
+    def run(self):
+        super().run()
+        import py_compile
+
+        import hy  # for compile hooks
+
+        for path in set(self.get_outputs()):
+            if path.endswith(".hy"):
+                py_compile.compile(
+                    path,
+                    invalidation_mode=py_compile.PycInvalidationMode.CHECKED_HASH,
+                )
+
+
+# both setup_requires and install_requires
+# since we need to compile .hy files during setup
+requires = [
+    'hy @ git+https://github.com/hylang/hy@master#egg=hy',
+]
+
+
 setuptools.setup(
     name = 'hyrule',
     version = None,
-    install_requires = [
-        'hy @ git+https://github.com/hylang/hy@master#egg=hy-1.0'],
+    setup_requires=['wheel'] + requires,
+    install_requires=requires,
     packages = setuptools.find_packages(),
-    package_data = {
-        'hyrule': ['*.hy', '__pycache__/*']},
+    package_data={'': ['*.hy']},
     author = "Paul Tagliamonte",
     author_email = "tag@pault.ag",
     description = 'A utility library for the Hy programming language',
@@ -33,4 +56,5 @@ setuptools.setup(
         "Programming Language :: Python :: 3.7",
         "Programming Language :: Python :: 3.8",
         "Programming Language :: Python :: 3.9",
-        "Programming Language :: Python :: 3.10"])
+        "Programming Language :: Python :: 3.10"],
+    cmdclass={'install': install})

